Client/Assets/WindowCloseHelper.js

```diff
@@ -3,9 +3,70 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 (function () {
-  var _close = window.close;
-  window.close = function () {
-    webkit.messageHandlers.windowCloseHelper.postMessage(null);
-    _close();
-  };
+    if (!window.__firefox__) {
+        window.__firefox__ = {};
+    }
+    if (!window.__firefox__.messages) {
+        window.__firefox__.messages = {};
+    }
+    window.__firefox__.messages.close = "FIREFOX_MESSAGE_CLOSE";
+ 
+    var _close = window.close;
+ 
+    window.close = function () {
+        webkit.messageHandlers.windowCloseHelper.postMessage(null);
+        _close();
+    };
+    // A close message applies to the current window, so we need to execute the
+    // close method in the context of the child window when we want to close a
+    // window opened via window.open(). To do this we use window.postMessage, at
+    // the expense of a side effect in the case of a website sending a message
+    // identical to one contained in window.__firefox__.messages.close.
+    window.addEventListener("message", function (event) {
+        if (event.data === window.__firefox__.messages.close) {
+            window.close();
+            event.stopPropagation();
+        }
+    }, true);
+    var _open = window.open;
+    window.open = function (strUrl, strWindowName, strWindowFeatures) {
+        // If we simply return the reference returned by window.open(),
+        // then (the read-only) window.close() will fail to execute when the
+        // opened window is cross-domain. Thus, we return a proxy object with an
+        // overwritten .close() method to invoke our special handle function.
+        opened = _open.apply(this, arguments);
+        // If we can overwrite properties, then cross-origin restrictions don't apply,
+        // so we can simply return the original window
+        var urlHost;
+        try {
+            urlHost = new URL(strUrl).host.split(".").slice(-2).join(".");
+        } catch (error) {
+            urlHost = null
+        }
+        var crossOrigin = window.location.host.split(".").slice(-2).join(".") !== urlHost;
```

Comment:
Can we just use `window.location.origin` here (and `url.origin` above)? Simpler, and would also catch different schemes/ports.